design/service-binding.md

```diff
@@ -0,0 +1,851 @@
+    - host: myhost2.example.com
+      port: 9092
+  - type: tls
+    addresses:
+    - host: myhost3.example.com
+      port: 9093
+  bootstrap:
+    plain: myhost1.example.com:9092,myhost2.example.com:9092
+    tls: myhost3.example.com
+```
+
+This has the advantage that the bootstrap servers information is at a known point in the `status` of the `Kafka` custom resource which makes is simple to annotate the CSV so the Service Binding Operator can find it.
+
+Because that structure is a little odd, perhaps this would be preferred.
+
+``` yaml
+bootstrap:
+- name: plain
+  value: myhost1.example.com:9092,myhost2.example.com:9092
+- name: tls
+  value: myhost3.example.com
+```
+
+This is trying of course to find a syntax that is sufficiently general that the Service Binding Operator can be extended to support it.
+
+``` yaml
+apiVersion: operators.coreos/com:v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: strimzi-cluster-operator.v0.16.2
+  namespace: placeholder
+  annotations:
+    containerImage: docker.io/strimzi/operator:0.16.2
+spec:
+  customresourcedefinitions:
+    owned:
+    - description: Represents a Kafka cluster
+      displayName: Kafka
+      kind: Kafka
+      name: kafkas.kafka.strimzi.io
+      version: v1beta1
+      statusDescriptors:
+      - description: The addresses of the bootstrap servers for Kafka clients
+        displayName: Bootstrap servers
+        path: status.bootstrap
+        x-descriptors:
+        - binding:env:object:secret:endpoints
+```
+
+#### Consuming client's ServiceBindingRequest
+
+All of the required annotations are applied to the `Kafka` CSV, so the binding should only need to refer to the `Kafka` CR.
+
+``` yaml
+apiVersion: service.binding/v1alpha1
+kind: ServiceBindingRequest
+metadata:
+  name: my-binding
+spec:
+  services:
+  - group: kafka.strimzi.io
+    kind: Kafka
+    version: v1beta1
+    resourceRef: my-cluster
+```
+
+The consuming client needs to know the listener name.
+
+The Service Binding Operator creates a `Secret` which contains:
+
+* **endpoints.<listener_name>** - bootstrap server information for all listeners
+
+
+### Option 2 - Generate a binding ConfigMap
+
+Rather than adding bootstrap server information to the `status`, Strimzi could create a ConfigMap and use the annotations to point the Service Binding Operator to that.
+
+Here's the content of the ConfigMap, with type and protocol added:
+
+``` yaml
+type=kafka
+protocol=kafka
+endpoints.plain=myhost1.example.com:9092,myhost2.example.com:9092
+endpoints.tls=myhost3.example.com
+```
+
+And the `status` changed to point to it:
+
+``` yaml
+apiVersion:kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+ ...
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: myhost1.example.com
+      port: 9092
+    - host: myhost2.example.com
+      port: 9092
+  - type: tls
+    addresses:
+    - host: myhost3.example.com
+      port: 9093
+  bindingConfigMap: binding-cm
```

Comment:
I think the advantage of this one is that you have the config map only for the service binding operator. That means it might be more flexible to change. The Status is an API which is harder to change (backwards compatibility, users might use the status pieces even outside of the service bindings). On the other hand if you change the content of the service binding CM to accomodate some service binding changes, it should be easier.

Reply:
But the status changes are minimal and not all specific to service binding. You can now just get Kafka bootstrap servers config in a more straightforward way.